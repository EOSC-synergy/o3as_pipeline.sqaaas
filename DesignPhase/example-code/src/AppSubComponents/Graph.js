import * as React from 'react';

/**
 * A containter for the displayed graph.
 * Displays the currently selected plot.
 */
export default function Graph() {
    return (<div>
        {/* Graph displayed with ApexCharts goes here */}
    </div>);
}