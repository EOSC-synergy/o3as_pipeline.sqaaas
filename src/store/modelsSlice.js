import { createSlice } from "@reduxjs/toolkit";

const STATISTICAL_VALUES = ["mean", "median", "derivative", "percentile"];

const initialState = {
    // currently active plot
    plotId: "tco3_zm",
    settings: {
        "tco3_zm": { 
            // this objects holds key-value-pairs, the keys being the model-group 
            // identifier and the values being the settings for each group 
            "all": { 
                // model group storing all information until it is possible 
                // to implement more model groups
                name: "All OCTS models",
                modelList: ["CCMI-1_ACCESS_ACCESS-CCM-refC2"],
                models: { // models is lookup table
                    "CCMI-1_ACCESS_ACCESS-CCM-refC2": { // single model
                        institute: "IMK",
                        dataset: {
                            x: [42],
                            y: [42],
                        },
                        color: "#000000",
                        plotStyle: "solid",
                        isVisible: true, // show/hide individual models from a group
                        mean: true,
                        derivative: true,
                        median: true,
                        percentile: true,
                    }
                },
                hidden: false, // show/hide complete group
                visibileSV: { // lookup table so the reducer impl. can be more convenient
                    mean: true,
                    derivative: true,
                    median: true,
                    percentile: true,
                }
            }
        },
        "tco3_return": {
            "all": { 
                name: "All Return/Recovery models",
                modelList: ["CCMI-1_ACCESS_ACCESS-CCM-refC2"],
                models: { // models is lookup table
                    "CCMI-1_ACCESS_ACCESS-CCM-refC2": { // single model
                        institute: "IMK",
                        dataset: {
                            x: [42],
                            y: [42],
                        },
                        hidden: false, // show/hide individual models from a group
                        mean: true,
                        derivative: true,
                        median: true,
                        percentile: true,
                        color: "#000000",
                        plotStyle: "solid",
                    }
                },
                hidden: false, // show/hide complete group
                visibileSV: { // lookup table so the reducer impl. can be more convenient
                    mean: true,
                    derivative: true,
                    median: true,
                    percentile: true,
                }
            }
        }
    }
}

const modelsSlice = createSlice({
    name: "models",
    initialState,
    reducers: {
        // set models
        // add models

        addModels(state, action) { 
            const { groupId, newModelList }  = action.payload;
            const activeSettings = state.settings[state.plotId];
            const selectedGroup = activeSettings[groupId];
            const activeModelList = selectedGroup.modelList;

            if (selectedGroup === undefined) {
                throw `Tried to access model-group with groupID "${groupId}" that is not yet defined!`;
            }

            // Add model to modelList if it is not already included
            for (var i = 0; i < newModelList.length; i++) {
                if (activeModelList.indexOf(newModelList[i]) !== -1) {
                    continue;
                }

                activeModelList.push(newModelList[i]);
            }

            // ---------------------------------------------------------
            // TODO: Add Model information to the model lookup table
            // ---------------------------------------------------------


        }, // expects groupID, list of models 
        removeModels(state, action) { 
            const { groupId, removeModelList }  = action.payload;
            const activeSettings = state.settings[state.plotId];
            const selectedGroup = activeSettings[groupId];
            const activeModelList = selectedGroup.modelList;

            if (selectedGroup === undefined) {
                throw `Tried to access model-group with groupID "${groupId}" that is not yet defined!`;
            }

            for (var i = 0; i < removeModelList.length; i++) {
                // Remove model from modelList
                activeModelList = activeModelList.filter(item => item !== removeModelList[i]);

                // Delete model object from models
                delete selectedGroup.models[removeModelList[i]];
            }

        }, // expects groupdID, list of models
        updatedModelGroup(state, action) { }, // implementing this requires more knowledge about the UI
        addedModelGroup(state, action) { },
        setVisibility(state, action) { 
            const { groupID, modelID, isVisible } = action.payload;
            const activeSettings = state.settings[state.plotId];
            const activeModel = activeSettings[groupID].models[modelID];
            if (activeModel === undefined) {
                throw `tried to access model with modelID "${modelID}" that is not yet defined!`;
            } 
            activeModel.isVisible = isVisible;
        },
        setStatisticalValuesIncluded(state, action) { // this is for an individual model
            // svType has to be a string of the form: 
            // mean | derivative | median | percentile
            const { groupID, modelID, svType, isIncluded } = action.payload;
            const activeSettings = state.settings[state.plotId];
            const activeModel = activeSettings[groupID].models[modelID];
            if (activeModel === undefined) {
                throw `tried to access model with modelID "${modelID}" that is not yet defined!`;
            }
            if (STATISTICAL_VALUES.includes(svType)) { // svType doesn't represent a valid statistical value
                throw `tried to set statistial value "${svType}" that is not a valid statistical value (${STATISTICAL_VALUES.join("|")})`;
            }
            activeModel[svType] = isIncluded;
        },
        setStatisticalValueForGroup(state, action) { // this is for an entire group
            const { groupID, svType, isIncluded } = action.payload;
            const activeSettings = state.settings[state.plotId];
            if (STATISTICAL_VALUES.includes(svType)) { // svType doesn't represent a valid statistical value
                throw `tried to set statistial value "${svType}" that is not a valid statistical value (${STATISTICAL_VALUES.join("|")})`;
            }
            activeSettings[groupID].visibileSV[svType] = isIncluded;
        },
    }
})

export const { 
    addModels,
    removeModels,
    updatedModelGroup,
    addedModelGroup,
    setVisibility, 
    setStatisticalValuesIncluded, 
    setStatisticalValueForGroup
} = modelsSlice.actions;

export default modelsSlice.reducer;

export const selectCurrentModelGroups = state => state.models[state.plot.plotId];
export const selectCurrentModelGroup = (state, groupId) => state.models[state.plot.plotId][groupId];