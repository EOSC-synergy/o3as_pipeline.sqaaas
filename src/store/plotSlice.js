import { createSlice } from "@reduxjs/toolkit"

/**
 * The initial state of the plotSlice defines the data structure in the 
 * store. Each plot has its own settings i.e. title, location etc.
 */
const initialState = {
    
    plotId: "tco3_zm", // currently active plot
    // maps plotids to their settings
    settings: {
        "tco3_zm": {
            name: "OCTS", // should show up in the drop down menu
            title: "OCTS Plot", // the title shown in the apexcharts generated chart
            location: {
                minLat: -90,
                maxLat: 90
            },
            displayXRange: {
                minX: 1960,
                maxX: 2100,
            },
            displayYRange: {
                minY: 200,
                maxY: 400,
            },
            months: [
                1
            ],
        },
        "tco3_return": {
            name: "Return/Recovery",
            title: "Return/Recovery Plot",
            location: { // custom user defined region
                minLat: -90,
                maxLat: 90
            },
            displayXRange: {
                minX: 0,
                maxX: 0,
            },
            displayYRange: {
                minY: 200,
                maxY: 400,
            },
            months: [
                1
            ],
        }
    }
}

/**
 * The plotSlice is generated by the redux toolkit. The reducers are defined here
 * and the corresponding actions are auto-generated.
 */
const plotSlice = createSlice({
    name: "plot",
    initialState,
    reducers: {
        setActivePlotId(state, action) { // e.g. dispatch(setActivePlotId({id: "tco3_zm"}))
            const { id } = action.payload
            state.plotId = id
        }, 
        setTitle(state, action) { // e.g. dispatch(setActivePlotId({title: "OCTS Plot"}))
            const { title } = action.payload
            state.settings[state.plotId].title = title
        },
        setLocation(state, action) { 
            const {minLat, maxLat} = action.payload
            const location = state.settings[state.plotId].location
            location.minLat = minLat
            location.maxLat = maxLat
        },
        setDisplayXRange(state, action) { 
            const {minX, maxX} = action.payload
            const displayXRange = state.settings[state.plotId].displayXRange
            displayXRange.minX = minX
            displayXRange.maxX = maxX
        },
        setDisplayYRange(state, action) { 
            const {minY, maxY} = action.payload
            const displayYRange = state.settings[state.plotId].displayYRange
            displayYRange.minY = minY
            displayYRange.maxY = maxY
        },
        setMonths(state, action) { 
            state.settings[state.plotId].months = action.payload.months
        },
    }   
})


/**
 * The here listed actions are exported and serve as an interface for
 * the view (our react components).
 */
export const {
    setActivePlotId,
    setTitle,
    setLocation,
    setDisplayXRange,
    setDisplayYRange,
    setMonths,
} = plotSlice.actions

/**
 * The reducer combining all reducers defined in the plot slice. 
 * This has to be included in the redux store, otherwise dispatching 
 * the above defined actions wouldn't trigger state updates.
 */
export default plotSlice.reducer

/**
 * This selectors allows components to select the current plot id
 * from the store. The plot id is a string using the same naming as the
 * o3as api e.g. tco3_zm or tco3_return
 * 
 * @param {object} state the global redux state
 * @returns {string} the current active plot id
 */
export const selectPlotId = state => state.plot.plotId

/**
 * This selectors allows components to select the current plot name
 * from the store. The plot name is the humand readable version specifying 
 * the plot e.g. OCTS or Return/Recovery
 * 
 * @param {object} state the global redux state
 * @returns {string} the current active plot name
 */
export const selectPlotName = state => state.plot.settings[state.plot.plotId].name

/**
 * This selectors allows components to select the current plot title
 * from the store.
 * 
 * @param {object} state the global redux state
 * @returns {string} the current active plot title
 */
export const selectPlotTitle = state => state.plot.settings[state.plot.plotId].title

/**
 * This selectors allows components to select the current plot location
 * from the store. The location is an object containg a minLat and maxLat attribute.
 * 
 * @param {object} state the global redux state
 * @returns {object} holds the current location that includes a minLat and maxLat attribute.  
 */
export const selectPlotLocation = state => state.plot.settings[state.plot.plotId].location

/**
 * This selectors allows components to select the current x range
 * from the store.
 * 
 * @param {object} state the global redux state
 * @returns {object} holds the current x range that includes minX and maxX
 */
export const selectPlotXRange = state => state.plot.settings[state.plot.plotId].displayXRange

/**
 * This selectors allows components to select the current y range
 * from the store.
 * 
 * @param {object} state the global redux state
 * @returns {object} holds the current x range that includes minY and maxY
 */

export const selectPlotYRange = state => state.plot.settings[state.plot.plotId].displayYRange

/**
 * This selectors allows components to select the current selected months
 * from the store.
 * 
 * @param {object} state the global redux state
 * @returns {array} array of strings describing the current selected months
 */

export const selectPlotMonths = state => state.plot.settings[state.plot.plotId].months
