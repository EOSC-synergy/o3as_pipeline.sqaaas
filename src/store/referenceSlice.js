import { createSlice } from "@reduxjs/toolkit"

const initialState = {
    "OCTS": {
        year: 1980,
        model: "CCMI-1_ACCESS_ACCESS-CCM-refC2",
        visibile: false,
        offsetApplied: false,
    }
}

const referenceSlice = createSlice({
    name: "reference",
    initialState,
    reducers: {
        setReferenceYear(state, action) {
            const {year} = action.payload;
            state.OCTS.year = year;
         },
        setReferenceModel(state, action) { 
            const {model} = action.payload;
            state.OCTS.model = model;
        },
        setVisibile(state, action) { },
        setOffsetApplied(state, action) { },
    }
})

export const {
    setReferenceYear,
    setReferenceModel,
    setVisibile,
    setOffsetApplied,
} = referenceSlice.actions

export default referenceSlice.reducer

export const selectCurrentReferenceSettings = state=> state.reference[state.plot.plotId]
