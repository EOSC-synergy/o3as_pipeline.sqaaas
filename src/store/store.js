import { configureStore } from '@reduxjs/toolkit'
import plotReducer from './plotSlice/plotSlice'
import modelsReducer from './modelsSlice/modelsSlice'
import referenceReducer from './referenceSlice'


export default configureStore({
  reducer: {
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
  }
})