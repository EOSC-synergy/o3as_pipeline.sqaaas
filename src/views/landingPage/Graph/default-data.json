[
  {
    "legalinfo": "https://o3as.data.kit.edu/policies/terms-of-use.html",
    "model": "CCMI-1_ACCESS_ACCESS-CCM-refC2",
    "plotstyle": {
      "color": "purple",
      "linestyle": "solid",
      "marker": "o"
    },
    "x": [
      "2000-01-16 00:00:00",
      "2001-01-16 00:00:00",
      "2002-01-16 00:00:00",
      "2003-01-16 00:00:00",
      "2004-01-16 00:00:00",
      "2005-01-16 00:00:00",
      "2006-01-16 00:00:00",
      "2007-01-16 00:00:00",
      "2008-01-16 00:00:00",
      "2009-01-16 00:00:00",
      "2010-01-16 00:00:00",
      "2011-01-16 00:00:00",
      "2012-01-16 00:00:00",
      "2013-01-16 00:00:00",
      "2014-01-16 00:00:00",
      "2015-01-16 00:00:00",
      "2016-01-16 00:00:00",
      "2017-01-16 00:00:00",
      "2018-01-16 00:00:00",
      "2019-01-16 00:00:00",
      "2020-01-16 00:00:00"
    ],
    "y": [
      333.5295715332031,
      343.16815185546875,
      332.8560485839844,
      338.5194396972656,
      331.2967834472656,
      330.7667541503906,
      312.93682861328125,
      314.9363098144531,
      328.24658203125,
      324.65325927734375,
      329.2712707519531,
      314.0989990234375,
      327.3851013183594,
      339.8885192871094,
      338.0392761230469,
      331.46929931640625,
      335.0693054199219,
      335.1351318359375,
      323.4892272949219,
      319.8822021484375
    ]
  }
]  