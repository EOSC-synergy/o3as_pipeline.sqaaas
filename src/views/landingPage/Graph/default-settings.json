{
    "options": {
        "xaxis": {
            "categories": []
        },
        "chart": {
            "id": "OCTS",
            "animations": {
                "enabled": false,
                "easing": "linear"
            },
            "toolbar": {
                "show": true,
                "tools":{
                    "download": true 
                }
            },
            "zoom": {
                "enabled": true,
                "type": "xy"
            }
        },
        "legend": {
            "show": false
        },
        "dataLabels": {
            "enabled": false
        },
        "tooltip": {
            "enabled": true,
            "shared": false
        },
        "colors": [],
        "stroke": {
            "width": [],
            "dashArray": []
        },
        "title": {
            "text": "OCTS Plot",
            "align": "center",
            "floating": false,
            "style": {
                "fontSize":  "30px",
                "fontWeight":  "bold",
                "fontFamily":  "undefined",
                "color":  "#263238"
            }
        }
        

    },
    "series": []
}