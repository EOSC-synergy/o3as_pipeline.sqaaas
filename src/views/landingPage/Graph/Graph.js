import * as React from 'react';
import {useSelector} from 'react-redux'
import { useGetRawPlotDataMutation } from '../../../services/API/apiSlice';
import { selectCurrentModelGroups } from '../../../store/modelsSlice';
import { selectCurrentPlotId, selectCurrentPlotType, selectCurrentSettings } from '../../../store/plotSlice';
import { selectCurrentReferenceSettings } from '../../../store/referenceSlice';
// import {calculatePlotSeries} from "../../../utils/math"

/**
 * A containter for the displayed graph.
 * Displays the currently selected plot.
 */
function Graph(props) {

    // const plotType = selectCurrentPlotType()
    // const plotId = selectCurrentPlotId()
    // const options = useSelector(state => selectCurrentSettings(state, plotId))
    // const models = useSelector(state => selectCurrentModelGroup(state, plotId, "all"))
    // const referenceSettings = useSelector(state => selectCurrentReferenceSettings(state, plotId))
    
    // const { data, isError, error, isLoading, isSuccess } = useGetRawPlotDataMutation(
    //     {
    //         plotType,
    //         latMin: options.latMin,
    //         latMax: options.latMax,
    //         modelList,
    //     }
    // )
    
    // if (isSuccess) {
    //     // updateCurrentModelGroup with fetched data
    // }


    // display loading spinner instead of graph until isSuccess = true
    // then load data
    
    // const seriesData = generatePlotSeries(models)

    return (<div>
        {/* Graph displayed with ApexCharts goes here */}
    </div>);
}

export default Graph;
