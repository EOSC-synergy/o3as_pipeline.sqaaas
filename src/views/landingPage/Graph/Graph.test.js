import Graph from './Graph'
import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import {createTestStore} from "../../../store/store"

let store;
beforeEach(() => {
  store = createTestStore();
}); 

jest.mock('react-apexcharts', () => {
    return {
        __esModule: true,
        default: () => {
            return <div />
        },
    }
})
jest.mock('apexcharts', () => ({ exec: jest.fn(() => { return new Promise((resolve) => { resolve("uri") }) }) }));

let store;
describe('tests graph component rendering', () => {
    beforeEach(() => {
        store = createTestStore();
    });

test.only('Renders OK', async () => {
    const { container } = render(<Provider store={store}><Graph /></Provider>);
    expect(container).toMatchSnapshot();
    it('Renders OK', async () => {
        const { container } = render(<Provider store={store}>
            <Graph />
        </Provider>
        );
        expect(container).toMatchSnapshot();
    });
});})