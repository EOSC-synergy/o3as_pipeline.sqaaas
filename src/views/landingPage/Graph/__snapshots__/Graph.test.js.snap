// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests graph component rendering renders with loading spinner if no data present 1`] = `
<div>
  <div
    style="display: flex; align-items: center; justify-content: center; height: 300px;"
  >
    <div
      class="spinner"
    >
      <h4>
        loading data
      </h4>
      <div
        class="loader"
        style="height: 8em; width: 8em;"
      />
    </div>
  </div>
</div>
`;
